### DISCUSSION TITLE
slow launch since 1.7.0?

### DISCUSSION THREAD

**Discussion by HarHarLinks** [davatorium/rofi#1499] [(Possible) Issues] [answered]
I have been using (lbonn wayland) rofi for a while but as my ~~daily~~ 5 minute driver since maybe mid 2020.
It has always been very snappy, I hit my launcher combo Mod4+D and start typing and it works.
That changed, I think, with the 1.7.0 update. Now I often find myself Hitting Mod4+D, starting to type and just catch myself before hitting enter on a chat or deleting the line back from my editor... all before the rofi window pops up.

I could find nobody sharing that exact experience, which is why I tried a bunch of things, including switching from lbonn to upstream rofi, playing with the cache settings, disabling theme, rewriting my Arc-Dark based theme using the new 1.7.0 features... nothing helps.

My hardware has not changed, and I am using sway on Arch Linux which I keep updated pretty much daily.

Any help would be appreciated.

**Comment by lbonn** [comment]
On the wayland fork, some slowdown could have been caused by the first clipboard implementation, before https://github.com/lbonn/rofi/pull/34 was merged. But if you can reproduce it on upstream rofi, it should be something else..?

Is the slowdown spurious or 100% reproducible?

**Comment by HarHarLinks** [comment, marked as answer]
It's reproducible, happens pretty much every time unless I close and instantly open it back up, which is why I tried to investigate the caching.
But it's the same on upstream rofi. I was using the aur/rofi-lbonn-wayland-git package which should have that PR merged for the last 10 days. the combination of these facts leads me to think that isn't it.

**Reply by DaveDavenport** [reply]
Can you get a timing report? https://github.com/davatorium/rofi/wiki/Debugging-Rofi#timing-traces
If you pastebin that it would give us a good starting point to debug this.

(I am very curious to see where the change is, no change from 1.6.x -> 1.7.0 springs to mind that would cause this.)

**Reply by HarHarLinks** [reply]
timing report: https://gist.github.com/HarHarLinks/e9c0c42d24ecf7e1b75ae23b54a02bca

**Reply by DaveDavenport** [reply]
It seems I/O is the biggest slow down, the reading and parsing of the desktop files:

```
(process:222281): Timings-DEBUG: 16:05:39.093: 0.035296 (0.009798): source/dialogs/run.c:get_apps:376 stop
(process:222281): Timings-DEBUG: 16:05:40.131: 1.073403 (1.038107): source/dialogs/drun.c:get_apps:1006 Get Desktop apps (start)
```

You can try using the `drun cache`

```bash
rofi -drun-use-desktop-cache -show drun
```

This should speed it up (after the cache is created).

Weird is, this has not really been changed  a lot between releases.

**Reply by DaveDavenport** [reply]
sorry misread that.. its between run and drun..   let me see
Odd place have nto seen that before.

Do you have plugins installed?  can you pastebin the output  of rofi -h  from "Monitor layout" point on?

**Reply by HarHarLinks** [reply]
> Do you have plugins installed?

I don't think so, how can I confirm that?

> can you pastebin the output of rofi -h from "Monitor layout" point on?

https://gist.github.com/HarHarLinks/27c1da95738e8f102878f70ab1c4c9a7

For reference, I call rofi from sway config like this: `bindsym $mod+d exec rofi -show drun` but the slowness is the same when running `rofi -show drun` from a shell so I would  exclude sway from the culprits.

**Reply by DaveDavenport** [reply]
thanks, that pastebin tells me what plugins there are installed (non outside of defaults). the timing trace is confusing  so sorry for the 300 questions:

Can you get a timing log of this:

```bash
G_MESSAGES_DEBUG=Timings rofi -show keys -modi keys -no-config
```

This should load the absolute minimum.
For me it launches and draws in 90ms on a old laptop:
```
(process:84609): Timings-DEBUG: 18:30:37.195: 0.087116 (0.000023): ../source/view.c:rofi_view_repaint:279 flush
```

**Reply by HarHarLinks** [reply]
This one seems fine!
https://gist.github.com/HarHarLinks/c89104ac70160f05b1ffe23b1d8fde5f

**Reply by DaveDavenport** [reply]
thanks.   so I suspect it is the run mode spidering the hdd.

Can you test and post this one:
```bash
G_MESSAGES_DEBUG=Timings rofi -show run -modi run -no-config
```
and this one:
```bash
G_MESSAGES_DEBUG=Timings rofi -show drun -modi drun -no-config
```

That should hopefully show the whole picture.

I am surprised that the 1.7.0 release caused the slow down, the modi itself have not changed significant.

**Reply by HarHarLinks** [reply]
>spidering the hdd.

It must have been a code change (in rofi or some underlying thing) then, as the hardware is the same SSD it has always been.

> ```shell
> G_MESSAGES_DEBUG=Timings rofi -show run -modi run -no-config
> ```
https://gist.github.com/HarHarLinks/d399bf9690be8492e99cab920504cdd3

> ```shell
> G_MESSAGES_DEBUG=Timings rofi -show drun -modi drun -no-config
> ```
https://gist.github.com/HarHarLinks/5d9f4e7a651e743a94a0cec5937af1a7

> I am surprised that the 1.7.0 release caused the slow down, the modi itself have not changed significant.

I might be the cause, it might be coincidental, I might also be misremembering and it started to happen only a week after or so. Either way thank you very much for helping me fix whatever causes it for me :)

**Reply by DaveDavenport** [reply]
The first one looks off. it does not have run mode.

But it looks like it must be drun mode.

**Reply by DaveDavenport** [reply]
Hopefully the last one, can you do:

```bash
G_MESSAGES_DEBUG=Timings rofi -show filebrowser -modi filebrowser -no-config
```

DRun (and run)  both create a file-browser for possible completion, that might cause the slow down (do you have tons of files in the default work directory of rofi).

If this is confirmed I will see to resolve this.

**Reply by HarHarLinks** [reply]
> The first one looks off. it does not have run mode.

I tried several times and couldn't repro with run mode then. However I retried just now and it was slow-ish again: https://gist.github.com/HarHarLinks/729ffd9ec45697072d6337478ca622d4

> ```shell
> G_MESSAGES_DEBUG=Timings rofi -show filebrowser -modi filebrowser -no-config
> ```

https://gist.github.com/HarHarLinks/936fcf750ea1e3f43465b05f3aec903b

> do you have tons of files in the default work directory of rofi

What would that directory be?
I don't have anything configured in these filebrowser settings, and we're testing with `-no-config` anyway.
https://github.com/davatorium/rofi/blob/1520a4a5cdd5debd0f829692f9e4147df2bde946/releasenotes/1.7.0/release-1.7.0.markdown#L56
```shell
$ ls -l ~ | wc -l
146
$ ls -l .config/rofi/ | wc -l
3
```

**Reply by DaveDavenport** [reply]
If not set in config, default is home directory,  if file-browser been used before it stores the last used directory.

**Reply by HarHarLinks** [reply]
I sometimes have a nextcloud mounted using webdav (davfs2) and a symlink from my ~ into that. That can naturally be slow to access.
However
- I also had this before rofi was slow
- I sometimes don't have it mounted and it is slow regardless (the symlink leads to a nonexistent place in that case)

**Reply by DaveDavenport** [reply]
I will change it so it won't initialize the completer on launch, that should hopefully fix it.
If it finds a symlink it does a stat on the actual link to figure out what it is, so I can see this being slow.
(I will also see if I can optimize that code a bit, already added one this morning).
Ill post here once I put that in git.

**Reply by DaveDavenport** [reply]
added a quick fix in git, where file-completer is only initialized when called.
Can you test git?

(btw to force it to be slow do  `echo 1 >  /proc/sys/vm/drop_caches`  as root to clear all caches).

**Reply by HarHarLinks** [reply]
I installed aur/rofi-git which is on `Version: 1.7.0-46-gccaaaf6a (makepkg)`.
testing with command: `sudo bash -c "echo 1 > /proc/sys/vm/drop_caches"; G_MESSAGES_DEBUG=Timings rofi -show drun -modi drun -no-config`
seems good!
`(process:94110): Timings-DEBUG: 14:30:59.739: 0.141039 (0.000024): ../rofi/source/view.c:rofi_view_repaint:279 flush`

**Reply by DaveDavenport** [reply]
great, thanks for testing.

**Reply by HarHarLinks** [reply]
thanks for helping and fixing, I'll mark as resolved!
